Extended permission model for image access

bbockelm opened the pull request:
This PR introduces two new features for _read-only_ containers (writable containers are unaffected):

1.  "Protected mode".  Allows access to images readable by a dedicated user account.   This has three possible settings
      * `none` (default): This feature is disabled.
      * `group`: The image must be readable by the dedicated user account *and* its group ownership must overlap with the invoking user's.  So, if user `bbockelm` is in group `foo` and `bar`, then the user would be able to access an image with mode `600` and ownership `singularity:foo`, `singularity:bar`, but not `singularity:baz`.
      * `user`: The image must be readable by the dedicated user account.  Given the user `bbockelm` above, they would be able to access an image mode `600` and ownership `singularity:baz` or mode `640` and ownership `root:singularity`.
      The dedicated user account is named `singularity` by default(when switching to this account, we set to the corresponding UID and primary GID).
2.  A boolean flag (`allow user image`) on whether users can provide their own images (defaults to `yes`).  When disabled, user-provided images (i.e., anything readable by the invoking user) are not allowed.

The default configuration values leave the permission model unchanged from the current code.

A few example use cases:
* Sysadmin only wants `singularity` to execute images she provides for the user.  In that case, set `allow user image = no` and `protected user mode = user`.
* Admin wants to provide a group-ownership based model for the users she maintains.  Users in group `physics` should not access the images available to group `chemistry`.  In that case, set `protected user mode = group`.
   * As a follow-up, we could put together a more fine-grained model that a user is only allowed to execute a fixed command inside the image (would prevent the user from doing arbitrary reads of the software inside the image).
* Admin wants only users belonging to a given group to create `singularity` images.  Set `allow user image = no` and `protected user mode = user`, and add these users to the `singularity` group.
   * In this case, users put their images in a publicly-accessible directory and allow other users to execute them.

bbockelm committed:
Add concept of 'protected mode'.

In this mode, one can execute images owned by the singularity user
under various file ownership conditions.

vsoch commented on etc/singularity.conf:
```diff
@@ -17,8 +17,30 @@ allow setuid = yes
+# Note that if 'allow user image' is no and 'protected image mode' is none,
+# then no image file may be used.
+allow user image = yes
+
```
ok, so this is sort of like a controller for extending permissions for singularity image? For example, by quickly changing the values in the config file (protected image mode to none and allow user image to no) then no containers can be used. This would be very useful in some kind of situation where singularity needed to be "frozen" but not uninstalled / removed. 

Are these permissions living / changing or do they apply to all images created after being set?

bbockelm replied:
The permissions examined are those of the image file.  Since they're enforced by Linux, it's based on whenever `singularity` is invoked.

vsoch commented on etc/singularity.conf:
```diff
@@ -17,8 +17,30 @@ allow setuid = yes
+# PROTECTED IMAGE MODE: [STRING]
+# DEFAULT: none
+# The 'protected image mode' allows users to start singularity inside images
+# that they do not own.  There are three allowed modes:
```
what do you mean start singularity inside images? Does that actually work? (reminds me of this ;P https://github.com/jpetazzo/dind)

vsoch commented on etc/singularity.conf:
```diff
@@ -141,3 +163,14 @@ container dir = /var/singularity/mnt
+# the 'singularity user'; this option allows admins to change the username we
+# use as a test.
+#
+# If user namespaces are enabled, the singularity user setting is ignored and
```
I don't understand this. Is singularity the user that owns the singularity process? What kind of tests would be done and why would admins want to change this? Why does it have to be just one user?

bbockelm replied:
`singularity` is the user that must own the image file (when utilizing protected image mode).  The test performed is literally whether `singularity` can `fopen()` the file (and whether it owns the image file).  I don't foresee admins changing the name of the user, but I dislike hardcoding usernames (what if the site already has a user named `singularity`, used for other purposes?).

Because we perform the checks as the `singularity` user (instead of looking at the mode / ownership and "recreating" POSIX), we:
* Are less likely to introduce bugs as we are using the Linux kernel's logic.
* Will function in the face of with less-common access control methods such as file attributes and ACLs.

Since there is the `group` mode that provides more fine-grained permissions, I don't think there's much use in allowing the admin to specify multiple `singularity` users.

bauerm97 replied:
@bbockelm What are your thoughts on allowing several `singularity user` lines? The use case for this might be letting user `mibauer` in group `science` bring an image into the cluster that anybody in group `science` could access.

bbockelm replied:
@bauerm97 -
Two options exist:
   1.  Simply keep the `allow user images = yes` default and allow user `mibauer` to use normal Unix permissions to share his images.
   2.  If `allow user images = no` is set for other reasons, you could add `mibauer` to the `singularity` group, set `protected image mode = user` and have him set those images to be owned by group `singularity`.
      * Note that this allows anyone on the host to access the image.

So (2) is close - but not exactly - to what you want.

I currently clear out the supplementary GID list when testing access from the `singularity` user.  I suppose we could keep these GIDS for the `singularity` user.  In that case, (2) would require that the image live in group-readable (not-world-readable) space and owned by the `singularity` group.

bbockelm committed:
Add tests for protected mode.

With noted exception where we don't correctly test for ownership
by user `singularity`, this should cover the new features.

bbockelm committed:
Force singularity user ownership in protected mode.

bauerm97 commented on src/lib/config_parser.c:
```diff
@@ -129,6 +129,11 @@ char *singularity_config_get_value(char *key) {
     return(NULL);
 }
 
+char *singularity_config_get_value_default(char *key, const char *def) {
```
Right now, default values are created by adding a line `key = default` into the configuration file that gets distributed. Are we sure we need to add this function? The first method also ensure that two different calls to the function will always return the same value. If we add this function, it would make sense to me that we should change every call to `config_get_value` -> `config_get_value_default`. Otherwise, simply adding the default entry into the config file should do. Agree?

bbockelm replied:
The problem is that we distribute a _config_ file, not a _defaults_ file.  That means, on upgrade, we have no clue as to whether our updated configs are in place.  For example, on RPM-based platforms, updated configs will be saved as `.rpmnew` if the admin changes _any_ line of the config file.  (We've already seen this occur.)

Hence, we're left with the unenviable task of making sure the code defaults match the config defaults (NOTE: other projects I have worked with solve this by either auto-generating the config file from the code defaults at build-time or simply removing defaults from the config file).

We already do this for all the boolean configuration knobs; this just appears the first time we've done it for strings.

Yes, the configuration system code really needs an uplift to make it future-proof.

bauerm97 replied:
In that case would it make sense to leave this out of the PR for now, and then go in and give the config system a bit of a facelift to account for all of this in a separate PR? That way we don't mix and match what we're solving in each PR and what we're doing in the code, but rather go through and systematically change what needs to be changed.

I could maybe handle that when I'm finished with the bootstrapping code. There's not much left with the bootstrapping code (just polish and design decisions at this point!!), but I am prioritizing my time on that at the moment.

bauerm97 commented on src/get-section.c:
```diff
@@ -64,7 +64,7 @@ int main(int argc, char ** argv) {
     }
 
-    singularity_message(DEBUG, "Iterating through file looking for sections matching: \%%s\n", section);
+    singularity_message(DEBUG, "Iterating through file looking for sections matching: %s\n", section);
```
Can I ask what the reasoning behind removing "%" is? It would make sense to include the % sign in the string so it reads out "%pre" or "%setup". I'm not sure if % is the correct way to escape a % sign, but I know that %% works. Maybe use that instead?

(This is not really important or imperative, just nitpicking)

bbockelm replied:
Ah - so that's what it was trying to attempt!  (As written, it was an invalid format specifier.)  In that case, it'll need to change to `%%%s`.

bbockelm committed:
Restore debug output that I had mauled.

gmkurtzer commented on src/lib/privilege.c:
```diff
@@ -156,6 +163,83 @@ void singularity_priv_escalate(void) {
+    sinfo.gid = pw->pw_gid;
+}
+
+void singularity_priv_escalate_singularity(void) {
```
This is about the biggest thing I was able to find... The second mention of 'singularity' seems to be the singularity user? singularity_priv_escalate_singularityuser() perhaps? In either case, this is not worth holding up the merge.

bbockelm replied:
Yeah - it's unfortunate naming, but it's long enough.  I guess we could do "escalate privilege to the singularity user"...

bauerm97 replied:
I know this has already been merged, but what do you think about giving singularity_priv_escalate() an input variable? Something like `singularity_priv_escalate(char *user)`. If `char *user` is NULL then we escalate to root, otherwise we escalate to the user specified by `char *user`.

@bbockelm @gmkurtzer

gmkurtzer replied:
That is an interesting idea! My first thought is that I like it, but I would need to go back and look at the logic in singularity_priv_escalate_singularity() and be sure we don't want to mix the code logic as it is nice having them somewhat discrete now. @bbockelm, what is your take?

bbockelm replied:
Well, I like having the API force the programmer to explicitly select one of the three relevant user states:
- The superuser.
- The "singularity user", which has special privileges for images.  Note that the singularity user is possibly distinct from the Unix user named `singularity`.
- The invoking user.

We can easily determine the programmer's intent - even via static analysis! - by looking at the function name.

`singularity_priv_escalate(char *user)` provides too much flexibility.  Stack corruption attacks could cause the semantics of the code (as `char *user` is in read-write memory while the function pointer is currently read-only) and hides the special semantics for the singularity user.

bbockelm replied:
(It's worth noting that we aren't _just_ switching users, but also doing special handling of primary and secondary group IDs.)
